### PR TITLE
Add RSxxxx diagnostic for compilation end diagnostic descriptors without CompilationEnd custom tag

### LINKED ISSUE #6282: Add RSxxxx diagnostic for compilation end diagnostic descriptors without CompilationEnd custom tag
Add a new RSxxxx diagnostic to our MS.CA.Analyzers package to flag descriptors being used to report a diagnostic that do not have the CompilationEnd custom tag: https://github.com/dotnet/roslyn/blob/440b6fc370f0ddcea68b445e183f1383c0d1bc6e/src/Compilers/Core/Portable/Diagnostic/WellKnownDiagnosticTags.cs#L56 custom tag 

See https://github.com/dotnet/roslyn/issues/65304#issuecomment-1309882140 for details

### PR DESCRIPTION
**Pull request by mavasani**
Closes #6282

### CODE REVIEW

**Comment by mavasani** [comment]
@Youssef1313 @Evangelink for review. Thanks!

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs [hunk 1]
```diff
@@ -299,6 +316,132 @@ public override void Initialize(AnalysisContext context)
...
                 }
 
+                // Flag descriptor fields that are used to report compilation end diagnostics,
+                // but do not have the required 'WellKnownDiagnosticTags.CompilationEnd' custom tag.
+                // See https://github.com/dotnet/roslyn-analyzers/issues/6282 for details.
+                if (compilationEndContextType.GetMembers(DiagnosticWellKnownNames.ReportDiagnosticName).FirstOrDefault() is IMethodSymbol compilationEndReportDiagnosticMethod)
+                {
+                    var diagnosticCreateMethods = diagnosticType.GetMembers("Create").OfType<IMethodSymbol>()
+                        .Where(m => m.IsPublic() && m.Parameters.Length > 0 && SymbolEqualityComparer.Default.Equals(m.Parameters[0].Type, diagnosticDescriptorType))
+                        .ToImmutableHashSet(SymbolEqualityComparer.Default);
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs:326]: This is 5 elements only. I'm not sure, but an array might have faster lookup than a hash set.

The difference might not be significant though.

**Review by Youssef1313 — COMMENTED** [review]
Does this work for `DiagnosticDescriptorHelper.Create` we have in this repo?

**Review by Youssef1313 — APPROVED** [review]
LGTM (given a quick look)

**Comment by mavasani** [comment]
> Does this work for `DiagnosticDescriptorHelper.Create` we have in this repo?

I think it should. Analyzing CustomTags argument is done by existing code in the analyzer. Even otherwise, I think we primarily care about analyzers outside this repo to benefit from this new warning.

**Review by Youssef1313 — COMMENTED** [review]
It looks like there is a violation in this repo?

https://github.com/dotnet/roslyn-analyzers/blob/e3784463e7cc21b6d81e0cc485bdbd66e5248a34/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/UseValidPlatformString.cs#L90-L91

https://github.com/dotnet/roslyn-analyzers/blob/e3784463e7cc21b6d81e0cc485bdbd66e5248a34/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/UseValidPlatformString.cs#L34-L59

Though this one can't easily be caught as `context.ReportDiagnostic` isn't invoked, but rather captured as a method group and passed down few calls to report a diagnostic.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 1 above)
Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs:326]: > The difference might not be significant though.

Agreed, I am going to leave it as-is for now.

**Comment by mavasani** [comment]
> It looks like there is a violation in this repo?
> 
> https://github.com/dotnet/roslyn-analyzers/blob/e3784463e7cc21b6d81e0cc485bdbd66e5248a34/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/UseValidPlatformString.cs#L90-L91
> 
> https://github.com/dotnet/roslyn-analyzers/blob/e3784463e7cc21b6d81e0cc485bdbd66e5248a34/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/UseValidPlatformString.cs#L34-L59
> 
> Though this one can't easily be caught as `context.ReportDiagnostic` isn't invoked, but rather captured as a method group and passed down few calls to report a diagnostic.

Thanks, I'll create a separate PR to fix this issue.

**Pull request merged**